p5rpc.modloader/Merging/SpdMerger.cs

```diff
@@ -0,0 +1,245 @@
+﻿using CriFs.V2.Hook.Interfaces;
+using FileEmulationFramework.Lib.Utilities;
+using PAK.Stream.Emulator.Interfaces;
+using Persona.Merger.Cache;
+using SPD.File.Emulator.Interfaces;
+using SPD.File.Emulator.Interfaces.Structures.IO;
+using static p5rpc.modloader.Merging.MergeUtils;
+
+namespace p5rpc.modloader.Merging;
+
+internal class SpdMerger : IFileMerger
+{
+    private readonly MergeUtils _utils;
+    private readonly Logger _logger;
+    private readonly MergedFileCache _mergedFileCache;
+    private readonly ICriFsRedirectorApi _criFsApi;
+    private readonly ISpdEmulator _spdEmulator;
+    private readonly IPakEmulator _pakEmulator;
+
+    internal SpdMerger(MergeUtils utils, Logger logger, MergedFileCache mergedFileCache, ICriFsRedirectorApi criFsApi, ISpdEmulator spdEmulator, IPakEmulator pakEmulator)
+    {
+        _utils = utils;
+        _logger = logger;
+        _mergedFileCache = mergedFileCache;
+        _criFsApi = criFsApi;
+        _spdEmulator = spdEmulator;
+        _pakEmulator = pakEmulator;
+    }
+
+    public void Merge(string[] cpks, ICriFsRedirectorApi.BindContext context)
+    {
+        // Variable setup
+        var input = _spdEmulator.GetEmulatorInput();
+        var pakGroups = _pakEmulator.GetEmulatorInput();
+        var pathToFileMap = context.RelativePathToFileMap;
+        var tasks = new List<ValueTask>();
+        Dictionary<string, List<string>> looseSpds = new();
+        Dictionary<string, PakSpdRoutes> pakSpdRoutes = new();
+        CachedFileSource[] cpkSources = cpks.Select(cpk => new CachedFileSource { LastWrite = File.GetLastWriteTime(cpk) }).ToArray();
+
+        foreach (RouteGroupTuple group in input)
+        {
+            var route = group.Route;
+            string routeDir = Path.GetDirectoryName(route) ?? "";
+
+            if (routeDir.Contains('.')) // Check for spds in a pak
+            {
+                foreach (var pakGroup in pakGroups)
+                {
+                    var looseSpdRoutes = pakGroup.Files.Files.Where(x => $@"{pakGroup.Route}\{x}".Contains(route, StringComparison.OrdinalIgnoreCase));
```

Comment:
Anyway, biggest change needed is here

```csharp
foreach (RouteGroupTuple group in input)
{
    var route = group.Route;
    string routeDir = Path.GetDirectoryName(route) ?? "";

    if (routeDir.Contains('.')) // Check for spds in a pak
    {
        foreach (var pakGroup in pakGroups)
        {
            var looseSpdRoutes = pakGroup.Files.Files.Where(x => $@"{pakGroup.Route}\{x}".Contains(route, StringComparison.OrdinalIgnoreCase));
            foreach (var spdRoute in looseSpdRoutes)
            {
```

This has 3 nested loops, O(N^3) so can blow up in terms of complexity, performance wise.

Assuming no nested PAKs, you can probably simplify this by:

# 1. Nabbing the name of the PAK from the route i.e.
- IndexOf `.pak` (rather than contains)
- Walk back to previous directory separator.
  - You can do this with `LastIndexOf` and slice of string ending at `.pak`.

# 2. Comparing against PAK name in `pakGroup`.

i.e. The name from emulator input.

If the two PAK names don't match, you can skip the inner code, resulting in a considerable speedup.